tensorflow/core/kernels/dml_deepcopy_op.cc

```diff
@@ -0,0 +1,80 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+Portions Copyright (c) Microsoft Corporation.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/core/common_runtime/dml/dml_operator_helper.h"
+#include "tensorflow/core/common_runtime/dml/dml_util.h"
+#include "tensorflow/core/framework/register_types.h"
+#include "tensorflow/core/kernels/dml_kernel_wrapper.h"
+#include "tensorflow/core/kernels/dml_ops_common.h"
+
+namespace tensorflow {
+
+class DmlDeepCopyKernel : public OpKernel {
+ public:
+  explicit DmlDeepCopyKernel(OpKernelConstruction* ctx) : OpKernel(ctx) {}
+
+  void Compute(OpKernelContext* ctx) override {
+    const Tensor& input = ctx->input(0);
+    const TensorShape& input_shape = input.shape();
+
+    Tensor* output = nullptr;
+    OP_REQUIRES_OK(ctx, ctx->allocate_output(0, input_shape, &output));
+
+    DmlDevice* device = static_cast<DmlDevice*>(ctx->device());
+    const auto& execution_context = device->GetExecutionContext();
+
+    D3D12BufferRegion input_buffer =
+        dml_util::CreateBufferForTensor(device, input);
+
+    D3D12BufferRegion output_buffer =
+        dml_util::CreateBufferForTensor(device, *output);
+
+    std::array<D3D12_RESOURCE_BARRIER, 2> barriers = {
```

Comment:
I believe this may be a race condition if two deep copies occur on the same input tensor. To make this safe, the transitions and copies should occur inside the same lock.

Reply:
This is actually safe, as the lifetime of the buffer is tied to the `D3D12BufferRegion`. While this method holds those objects, nobody else can re-use those buffers.

---
In reply to: [481370801](https://github.com/microsoft/tensorflow-directml/pull/4#discussion_r481370801) [](ancestors = 481370801)